File: demojava/src/main/java/com/teskalabs/seacat/demojava/MainActivity.java
package com.teskalabs.seacat.demojava;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.TextView;

import com.teskalabs.seacat.SeaCat;

import java.io.BufferedReader;
import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.MalformedURLException;
import java.net.URL;
import java.security.cert.X509Certificate;
import java.text.SimpleDateFormat;

import javax.net.ssl.HttpsURLConnection;

public class MainActivity extends AppCompatActivity {

    TextView identityTV;
    TextView sequenceTV;
    TextView validFromTV;
    TextView validToTV;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        identityTV = findViewById(R.id.identityTV);
        sequenceTV = findViewById(R.id.sequenceTV);
        validFromTV = findViewById(R.id.validFromTV);
        validToTV = findViewById(R.id.validToTV);
    }

    @Override
    protected void onStart() {
        super.onStart();
        update();


    }

    private void update() {
        SeaCat seacat = SeaCat.getInstance();

        identityTV.setText(String.format("Identity: %s", seacat.getIdentity().toString()));

        X509Certificate cert = seacat.getIdentity().getCertificate();

        if (cert != null) {
            SimpleDateFormat format = new SimpleDateFormat("dd.MMM yyyy HH:mm:ss");

            sequenceTV.setText("Sequence: %s".format(cert.getSerialNumber().toString()));
            validFromTV.setText("Valid from %s".format(format.format(cert.getNotBefore())));
            validToTV.setText("Valid to %s".format(format.format(cert.getNotAfter())));

        } else {
            sequenceTV.setText("Sequence -");
            validFromTV.setText("Valid from -");
            validToTV.setText("Valid to -");
        }

    }

    public void onRestCallClicked(View view) {
        final SeaCat seacat = SeaCat.getInstance();

        Thread thread = new Thread() {

            @Override
            public void run() {
                try {
                    URL url = new URL("https://zscanner.seacat.io/medicalc/v3.1/departments");

                    HttpsURLConnection connection = (HttpsURLConnection) url.openConnection();

                    // Configure the HTTPS connection to use SeaCat SSL context
                    connection.setSSLSocketFactory(seacat.getSslContext().getSocketFactory());

                    ByteArrayOutputStream buffer = new ByteArrayOutputStream();
                    InputStream is = connection.getInputStream();
                    int nRead;
                    byte[] data = new byte[1024];
                    while ((nRead = is.read(data, 0, data.length)) != -1) {
                        buffer.write(data, 0, nRead);
                    }

                    buffer.flush();
                    byte[] bytes = buffer.toByteArray();

                    StringBuilder result = new StringBuilder();
                    for (byte b : bytes) {
                        result.append(String.format("%02X", b));
                    }
                    Log.i("MainActivity", "Downloaded: " + result);


                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        };
        thread.start();

    }
}

File: demo/src/main/java/com/teskalabs/seacat/demo/MainActivity.kt
package com.teskalabs.seacat.demo

import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.teskalabs.seacat.SeaCat
import kotlinx.android.synthetic.main.activity_main.*
import java.net.URL
import java.text.SimpleDateFormat
import javax.net.ssl.HttpsURLConnection


class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
    }


    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return super.onCreateOptionsMenu(menu)
    }

    override fun onStart() {
        super.onStart()
        update()
    }

    override fun onStop() {
        super.onStop()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {

            R.id.action_revoke_identity -> {
                SeaCat.identity.revoke()
                Toast.makeText(applicationContext, "Identity has been revoked!", Toast.LENGTH_LONG).show()
                startActivity(Intent(this@MainActivity, SplashActivity::class.java))
                finish()
                return true
            }

            R.id.action_renew_identity -> {
                SeaCat.identity.renew()
            }
        }

        return super.onOptionsItemSelected(item)
    }

    private fun update() {
        identityTV.text = "Identity: %s".format(SeaCat.identity.toString())

        val cert= SeaCat.identity.certificate
        if (cert != null) {
            val format = SimpleDateFormat("dd.MMM yyyy HH:mm:ss")

            sequenceTV.text = "Sequence: %s".format(cert.serialNumber.toString())
            validFromTV.text = "Valid from %s".format(format.format(cert.notBefore))
            validToTV.text = "Valid to %s".format(format.format(cert.notAfter))

        } else {
            sequenceTV.text = "Sequence -"
            validFromTV.text = "Valid from -"
            validToTV.text = "Valid to -"
        }

        if (SeaCat.identity.isInsideSecureHardware) {
            insideHardwareTV.text = "Key stored in hardware :-)"
            insideHardwareTV.setTextColor(Color.GREEN)
        } else {
            insideHardwareTV.text = "Key NOT stored in hardware :-("
            insideHardwareTV.setTextColor(Color.RED)
        }
    }

    fun onRestCallClicked(view: android.view.View) {
        Thread({
            val url = URL("https://seacat-demo.seacat.io/hello")

            val connection = url.openConnection() as HttpsURLConnection
            // Configure the HTTPS connection to use SeaCat SSL context
            connection.setSSLSocketFactory(SeaCat.sslContext.socketFactory)

            val istream = connection.getInputStream()
            val response_body = String(istream.readBytes())

            Log.i(SeaCat.TAG, "Downloaded: " + response_body)
        }).start()
    }

}
